### PR TITLE
[Relay] Continuation Passing Style

### PR DESCRIPTION
**Pull request by MarisaKirisame**
The code implement a higher order continuation passing style transform in the spirit of (http://matt.might.net/articles/cps-conversion/, the higher order transform section).

This help training by helping PE which help AD.

Suppose we have some code, of the form
> let a = if b then c else d in e

in the partial evaluator, if b is dynamic c and d will be partially evaluated, and their result will be thrown away, so it cannot help e.

However, if we cps the program, it will become
> let f a = e in if b then f c else f d

with e inlined, and possible to do further simplification base on the partial evaluator result of c/d.
There are two big design choice this commit to:

- polymorphic continuation because we want to type check things. this mean we move from HM to System F.
- CPS itself. It is somewhat un-revertable, and has implication on further optimization. Some say it is good while some say it is bad.

We should discuss a bit.
@jroesch @tqchen @nhynes

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
@junrushao1994 @vinx13 I got it print what you guys want.

**Comment by MarisaKirisame** [comment]
turn out system f is not needed.

**Comment by vinx13** [comment]
Nice work! I will check if this is helpful for more complex cases of AD

**Comment by vinx13** [comment]
Is global var not yet supported?

**Comment by MarisaKirisame** [comment]
@vinx13 it is not supported, but there is no difficulty (only that I havent write some C++).
there are integration test for CPS, and CPS is itself a bit hard to test (the property) alone, as this one does a bit of simplification to remove immediate continuation.

File: src/relay/pass/to_cps.cc [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+      });
+    }
+
+    Expr VisitExpr_(const CallNode* op, const MCont& k) final {
+      if (op->op.as<OpNode>() || op->op.as<ConstructorNode>()) {
+        tvm::Array<Expr> args;
+        std::function<Expr()> next;
+        next = [&]() {
+          if (args.size() == op->args.size()) {
+            return LetList::Let(CallNode::make(op->op, args), k);
```
Comment (vinx13) [inline src/relay/pass/to_cps.cc]: `CallNode::make(op->op, args, op->attrs, op->type_args)` otherwise you lose attributes

**Comment by MarisaKirisame** [comment]
This is ready to review.
@zhiics I cant make UnCPS a transform as it only work on certain type.
@icemelon9 @wweic @srkreddy1238 can you guys also help review?

**Comment by tqchen** [comment]
One thing @MarisaKirisame given that we are moving towards new transform.h and Module->Module, would be great to keep everything in the new style

**Comment by MarisaKirisame** [comment]
@tqchen I am unable to make UnCPS into the new style as it only work on some of the function. We probably want to discuss design.

**Comment by tqchen** [comment]
I think we might want to have a conversation about where should we put the remaining function->function transformations into.

**Pull request merged**